app/models/collection.rb

```diff
@@ -162,6 +162,54 @@ def has_untranscribed_pages?
     self.works.joins(:work_statistic).where('work_statistics.total_pages > work_statistics.transcribed_pages').exists?
   end
 
+  def update_works_stats
+    works = self.works.includes(:work_statistic)
+    works_stats = get_works_stats_hash
+    works.each do |w|
+      w.work_statistic.recalculate_from_hash(works_stats[w.id]) if works_stats.key?(w.id)
```

Comment:
What happens to work statistics if there is no work_stats hash containing the work key?  Are there any conditions under which that might happen?

Reply:
One option might be to have `Work#recalculate_from_hash` simply test for `nil` in its argument and fall-back to recalculating from the database.

Reply:
I think that guard statement `is meant only to prevent any major catastrophes in the event that (somehow) the key doesn't exist. It  *should* in every case. That said, I think your approach is better and makes the `recalculate_from_hash` a more usable helper function. For example, as you suggest:
```
def recalculate_from_hash(stats=nil)
    stats = get_stats_hash if stats.nil?

    self[:total_pages] = stats[:total]
...
```
Also allows the old `recalculate` function to go from:
```
def recalculate(_options = {})
    stats = get_stats_hash
    recalculate_from_hash(stats)
    recalculate_parent_statistics
  end

```
to
```
  def recalculate(_options = {})
    recalculate_from_hash
    recalculate_parent_statistics
  end
```
And, of course, makes the guard statement unnecessary.

I've lightly refactored `get_works_stats_hash`, however,  to make the explicit mapping between the `works` and the `stats_hash` by passing in a list of `ids` rather than generating this list within the method's body. Thus, by definition, the keys in the stats hash are those of the works list.